File: gui_view/src/main/java/com/goexp/galgame/gui/view/game/listview/simplelist/small/HeaderController.java
package com.goexp.galgame.gui.view.game.listview.simplelist.small;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.util.Tags;
import com.goexp.galgame.gui.util.res.Images;
import com.goexp.galgame.gui.util.res.LocalRes;
import com.goexp.galgame.gui.view.DefaultController;
import com.goexp.galgame.gui.view.common.jump.JumpBrandController;
import com.goexp.galgame.gui.view.game.HomeController;
import com.goexp.galgame.gui.view.game.detailview.part.DateShowController;
import com.goexp.galgame.gui.view.game.part.StateChangeController;
import javafx.fxml.FXML;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Label;
import javafx.scene.effect.ColorAdjust;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.HBox;
import javafx.scene.text.Text;

public class HeaderController extends DefaultController {

    @FXML
    private DateShowController dateviewController;

    @FXML
    private Hyperlink linkView;


    @FXML
    private JumpBrandController brandJumpController;

    @FXML
    private StateChangeController changeStateController;

    @FXML
    private ImageView imageImg;

    @FXML
    private HBox boxStar;

    @FXML
    private Text txtName;

    @FXML
    private Text txtSubName;


    @FXML
    private HBox boxTag;

    private Game targetGame;

    protected void initialize() {

        linkView.setOnAction((e) -> {

            HomeController.$this.loadDetail(targetGame);
        });

    }


    public void load(Game game) {


        loadWithoutImage(game);

        if (game.smallImg != null && game.smallImg.startsWith("http")) {
            imageImg.setImage(Images.GameImage.tiny(game));

        } else {
            imageImg.setImage(null);
        }


    }


    public void setImage(Image image) {
        imageImg.setImage(image);
    }

    private void loadWithoutImage(Game game) {
        this.targetGame = game;


        //        webjumpController.load(game);
        changeStateController.load(game);
        brandJumpController.load(game.brand);
        //        starChangeController.load(game);


        txtName.setText(game.getMainName());
        txtSubName.setText(game.getSubName());

        if (game.tag.size() > 0) {

            boxTag.getChildren().setAll(Tags.toNodes(game.tag, str -> {
                var tagLabel = new Label(str);

                tagLabel.getStyleClass().add("tag");
                tagLabel.getStyleClass().add("tagbig");

                return tagLabel;
            }));
        } else {
            boxTag.getChildren().clear();
        }

        dateviewController.load(game.publishDate);

        boxStar.getChildren().clear();
        var image = LocalRes.HEART_16_PNG.get();
        for (var i = 0; i < game.star; i++) {
            boxStar.getChildren().add(new ImageView(image));
        }

        if (game.state.get().value <= GameState.BLOCK.value)
            imageImg.setEffect(new ColorAdjust(0, -1, 0, 0));
        else
            imageImg.setEffect(null);

    }
}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/listview/simplelist/small/HeaderController.scala
package com.goexp.galgame.gui.view.game.listview.simplelist.small

import com.goexp.common.util.string.ConsoleColors.RED
import com.goexp.galgame.common.model.game.GameState
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.util.Tags
import com.goexp.galgame.gui.util.res.LocalRes
import com.goexp.galgame.gui.util.res.gameimg.GameImage
import com.goexp.galgame.gui.view.common.jump.JumpBrandController
import com.goexp.galgame.gui.view.game.detailview.part.DateShowController
import com.goexp.galgame.gui.view.game.part.StateChangeController
import com.goexp.galgame.gui.view.{DefaultController, MainController}
import javafx.fxml.FXML
import javafx.scene.control.{Hyperlink, Label}
import javafx.scene.effect.ColorAdjust
import javafx.scene.image.{Image, ImageView}
import javafx.scene.layout.HBox
import javafx.scene.text.Text

class HeaderController extends DefaultController {
  @FXML private var dateviewController: DateShowController = _
  @FXML private var brandJumpController: JumpBrandController = _
  @FXML private var changeStateController: StateChangeController = _


  @FXML private var linkView: Hyperlink = _
  @FXML private var imageImg: ImageView = _
  @FXML private var boxStar: HBox = _
  @FXML private var txtName: Text = _
  @FXML private var txtSubName: Text = _
  @FXML private var boxTag: HBox = _
  private var targetGame: Game = _

  override protected def initialize() = {
    linkView.setOnAction(_ => MainController().loadDetail(targetGame))

  }

  def load(game: Game) = {
    logger.debug(s"Game[${RED.s(game.id.toString)}] ${RED.s(game.name)} Date:${game.publishDate} img:${game.smallImg}  state:<${
      Option(game.state).map {
        _.get
      }.getOrElse("--")
    }>")


    loadWithoutImage(game)
    if (game.isOkImg)
      imageImg.setImage(new GameImage(game).tiny)
    else
      imageImg.setImage(null)
  }

  def setImage(image: Image) = imageImg.setImage(image)

  private def loadWithoutImage(game: Game) = {

    this.targetGame = game
    changeStateController.load(game)
    brandJumpController.load(game.brand)
    val titles = game.getTitles
    txtName.setText(titles.mainTitle)
    txtSubName.setText(titles.subTitle)

    if (game.tag.size > 0)
      boxTag.getChildren.setAll(
        Tags.toNodes(game.tag, (str: String) => {
          val tagLabel = new Label(str)
          tagLabel.getStyleClass.add("tag")
          tagLabel.getStyleClass.add("tagbig")
          tagLabel
        }))
    else
      boxTag.getChildren.clear()

    dateviewController.load(game.publishDate)

    val image = LocalRes.HEART_16_PNG

    boxStar.getChildren.clear()
    val stars = Range(0, game.star).to(LazyList).map { _ => new ImageView(image) }.toArray
    boxStar.getChildren.addAll(stars: _*)


    if ((game.state.get eq GameState.BLOCK) || (game.state.get eq GameState.SAME))
      imageImg.setEffect(new ColorAdjust(0, -1, 0, 0))
    else
      imageImg.setEffect(null)
  }
}